### PR TITLE
O2bert + apex pipeline functions

### PR DESCRIPTION
**Pull request by shanmugamr1992**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 1]
```diff
@@ -112,25 +164,58 @@ def forward(self, input_ids, attention_mask, token_type_ids, lm_labels=None):
...
+            forward_only=False,
+            tensor_shape=tensor_shape,
+            dtype=self.autocast_dtype,
+            grad_scaler=self.trainer.precision_plugin.scaler if self.cfg.precision == 16 else None,
+            custom_sync_context_handler=None,
+        )
+
+        loss_tensors_list = [loss_reduced['avg'] for loss_reduced in losses_reduced_per_micro_batch]
+        loss_tensor = torch.vstack(loss_tensors_list)
+        loss_mean = loss_tensor.mean(axis=0)
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:191]: The loss_mean doesn't seem to be averaged across data parallel group. Could you confirm where it is averaged?

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 2]
```diff
@@ -152,26 +237,56 @@ def training_step(self, batch, batch_idx):
...
+            tensor_shape=tensor_shape,
+            dtype=self.autocast_dtype,
+        )
+        loss_tensors_list = [loss_reduced['avg'] for loss_reduced in losses_reduced_per_micro_batch]
+        loss_tensor = torch.vstack(loss_tensors_list)
+        loss_mean = loss_tensor.mean(axis=0)
+        return loss_mean[0]
 
     def validation_epoch_end(self, outputs):
         averaged_loss = torch.stack(outputs).mean()
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:288]: same comments as training loss average. where is the average across data parallel group happening? @ericharper , I notice there is no explicit average in the GPT model code. Could you confirm?

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 3]
```diff
@@ -212,7 +327,9 @@ def process_batch(self, batch):
         datatype = torch.int64
 
         data = batch
-        data_b = tensor_parallel.broadcast_data(keys, data, datatype)
+        data_b = tensor_parallel.broadcast_data(
+            keys, data, datatype
+        )  # IS THIS NEEDED (NOT PRESENT IN GPT3 BUT PRESENT IN BERT NO-BATCH IMPLEMENTATION)
```
Comment (yidong72) [inline nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:332]: It is not needed. data is distributed for all the workers correctly. no need of broadcast.

**Review by yidong72 — COMMENTED** [review]
If you just want to add O2, why do you have to use the apex `forward_backward_no_pipelining`, which is an overkill unless it is intended to implement the pipeline parallel in the future. I can enable O2 for the RETRO model without using it, just fewer lines of code changes. Also do you want to enable `sequence_parallel`? The current implementation seems to be broken. 
Lastly, could you add the ci tests to make sure O2 and/or sequence parallel is working properly.

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 4]
```diff
@@ -112,25 +164,58 @@ def forward(self, input_ids, attention_mask, token_type_ids, lm_labels=None):
...
-            reduced_loss = average_losses_across_data_parallel_group([loss, lm_loss, sop_loss])
-            self._reduced_loss_buffer.append(reduced_loss[0])
-            self._reduced_lm_loss_buffer.append(reduced_loss[1])
-            self._reduced_sop_loss_buffer.append(reduced_loss[2])
+
+        self._optimizer.zero_grad()  # DO WE NEED THIS
+        batch_for_pipeline = self.process_batch(batch)
+        tensor_shape = [self.cfg.encoder_seq_length, self.cfg.micro_batch_size, self.cfg.hidden_size]
+
+        if self.megatron_amp_o2 and not self.cfg.get('sequence_parallel', False):
```
[the remark this answers is not in the document]
Comment (shanmugamr1992) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:172]: Nope. Will get rid of it. We will be implementing pipeline parallel as well soon. 
CI tests will add .

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 5]
```diff
@@ -112,25 +164,58 @@ def forward(self, input_ids, attention_mask, token_type_ids, lm_labels=None):
...
+
+        losses_reduced_per_micro_batch = forward_backward_no_pipelining(
+            forward_step_func=self.get_forward_output_and_loss_func(),
+            batch=batch_for_pipeline,
+            model=self.model,
+            forward_only=False,
+            tensor_shape=tensor_shape,
+            dtype=self.autocast_dtype,
+            grad_scaler=self.trainer.precision_plugin.scaler if self.cfg.precision == 16 else None,
+            custom_sync_context_handler=None,
```
[the remark this answers is not in the document]
Comment (shanmugamr1992) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:186]: Nope. That wont be there for now.

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py (hunk 1 above)
Comment (shanmugamr1992) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:191]: It happens in the loss_func, inside get_forward_output_and_loss_func

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shanmugamr1992) [reply nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:172]: @MaximumEntropy  We could actually skip the apex part of it if we are not doing pipeline parallel right ? Cant we just directly call the model  as Yi said?

File: nemo/collections/nlp/models/language_modeling/megatron_bert_model.py [hunk 6]
```diff
@@ -91,8 +104,48 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
             add_binary_head=cfg.bert_binary_head,
             megatron_legacy=cfg.get('megatron_legacy', False),
         )
-        # not using amp o2
-        self.megatron_amp_o2 = False
+
+        if self.megatron_amp_o2:
+            if isinstance(self.model, list):
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_bert_model.py:109]: Can we remove the conditions that check for whether a model is a list? This should only happen for interleaved pipeline parallelism which is only supported for GPT. You can take a look at how this bit is implemented for T5.

**Review by MaximumEntropy — COMMENTED** [review]
LGTM! Just a minor comment. Will let @yidong72 give the final approval.

**Review by yidong72 — COMMENTED** [review]
Loos good. could you add the CI test to cover the O2 pretraining case?

**Review by yidong72 — DISMISSED** [review]
LGTM

**Review by MaximumEntropy — APPROVED** [review]
